### PR TITLE
Add initial tests for user challenge management REST API

### PR DESCRIPTION
**Pull request by ayshsandu**
- enable supporting SwaggerValidationFilter along with rest-assured

### CODE REVIEW

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    @BeforeMethod(alwaysRun = true)
+    public void testInit() {
+
+        RestAssured.basePath = basePath;
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void testFinish() {
+
+        RestAssured.basePath = "";
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java]: Shall we use StringUtils.EMPTY ?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/common/RESTAPIUserTestBase.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    public static final String API_USERS_BASE_PATH_WITH_TENANT_CONTEXT = "/t/%s" + API_USERS_BASE_PATH;
+
+    protected void testInit(String apiVersion, String apiPackageName, String apiDefinitionName, String tenantDomain)
+            throws IOException, XPathExpressionException {
+
+        String basePathInSwagger = String.format(API_USERS_BASE_PATH, apiVersion);
+        String basePath = String.format(API_USERS_BASE_PATH_WITH_TENANT_CONTEXT,
+                tenantDomain, apiVersion);
+        super.init(apiPackageName, apiDefinitionName, basePathInSwagger, basePath);
+
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/common/RESTAPIUserTestBase.java]: Additional empty line.

File: modules/integration/tests-integration/tests-backend/src/test/resources/org/wso2/identity/integration/test/rest/api/user/challenge/v1/update-challenge-answers-correct-1.json [hunk 3]
```diff
@@ -0,0 +1,9 @@
+{
+    "challengeQuestion": {
+      "locale": "en_US",
+      "question": "Favorite vacation location ?",
+      "questionId": "question4"
+
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/resources/org/wso2/identity/integration/test/rest/api/user/challenge/v1/update-challenge-answers-correct-1.json]: Additional empty line.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                .header(HttpHeaders.ACCEPT, ContentType.JSON)
+                .log().all()
+                .filter(validationFilter)
+                .when()
+                .get("/me/challenge-answers")
+                .then()
+                .assertThat()
+                .statusCode(200);
+    }
+
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java]: Additional empty line.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            int character;
+            while ((character = bufferedInputStream.read()) != -1) {
+                char value = (char) character;
+                resourceFile.append(value);
+            }
+
+            return resourceFile.toString();
+        }
+    }
+
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java]: Additional empty line.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        this.basePath = basePath;
+        String yaml = getAPISwaggerDefinition(apiPackageJar, swaggerDefinition, basePathInSwagger, basePath);
+        swaggerRequestResponseValidator = SwaggerRequestResponseValidator
+                .createFor(yaml)
+                .build();
+        validationFilter = new SwaggerValidationFilter(swaggerRequestResponseValidator);
+    }
+
+    protected void conclude() {
+        RestAssured.basePath = "";
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java]: Shall we use StringUtils.EMPTY ?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+    private static final String API_WEB_APP_ROOT = File.separator + "repository"
+            + File.separator + "deployment" + File.separator + "server" + File.separator + "webapps" + File
+            .separator + "api" + File.separator + "WEB-INF" + File.separator
+            + "lib" + File.separator;
+    private static final String JAR_EXTENSION = ".jar";
+    private static final String SERVICES = "/services";
+
+    protected AutomationContext context;
+    protected String basePath = "";
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java]: Shall we use StringUtils.EMPTY ?

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.stream.Collectors;
+import javax.xml.xpath.XPathExpressionException;
+
+public abstract class RESTTestBase {
```
Comment (janakamarasena) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java]: Shall we add class level comment

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java (hunk 5 above)
Comment (ayshsandu) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java (hunk 1 above)
Comment (ayshsandu) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/common/RESTAPIUserTestBase.java (hunk 2 above)
Comment (ayshsandu) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/common/RESTAPIUserTestBase.java]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/resources/org/wso2/identity/integration/test/rest/api/user/challenge/v1/update-challenge-answers-correct-1.json (hunk 3 above)
Comment (ayshsandu) [reply modules/integration/tests-integration/tests-backend/src/test/resources/org/wso2/identity/integration/test/rest/api/user/challenge/v1/update-challenge-answers-correct-1.json]: fixed

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+     *
+     * @param apiPackageJar
+     * @param swaggerDefinition
+     * @param basePathInSwagger
+     * @param basePath
+     * @throws IOException
+     * @throws XPathExpressionException
+     */
+    protected void init(String apiPackageJar, String swaggerDefinition, String basePathInSwagger, String basePath)
+            throws IOException, XPathExpressionException {
```
Comment (omindu) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/common/RESTTestBase.java]:
```suggestion
            throws IOException, XPathExpressionException {
            
```

**Comment by ayshsandu** [comment]
Merge only after https://github.com/wso2/identity-api-user/pull/32/

File: modules/integration/tests-integration/tests-backend/src/test/resources/org/wso2/identity/integration/test/rest/api/user/challenge/v1/challenge-answer-set1-q4-correct.json [hunk 10]
```diff
@@ -0,0 +1,8 @@
+{
+    "challengeQuestion": {
+      "locale": "en_US",
+      "question": "Favorite vacation location ?",
+      "questionId": "question4"
+    },
+    "answer": "Paris"
+  }
```
Comment (omindu) [inline modules/integration/tests-integration/tests-backend/src/test/resources/org/wso2/identity/integration/test/rest/api/user/challenge/v1/challenge-answer-set1-q4-correct.json]:
```suggestion
 }
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    static {
+        try {
+            swaggerDefinition = getAPISwaggerDefinition(API_PACKAGE_NAME, API_DEFINITION_NAME);
+        } catch (IOException e) {
+            Assert.fail(String.format("Unable to read the swagger definition %s from %s", API_DEFINITION_NAME,
+                    API_PACKAGE_NAME), e);
+        }
+    }
+
+    void initUrls(String pathParam) {
```
Comment (omindu) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java]:
```suggestion
    private void initUrls(String pathParam) {
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserMeNegativeTest.java [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+import org.wso2.carbon.CarbonConstants;
+import org.wso2.carbon.automation.engine.context.TestUserMode;
+import org.wso2.carbon.um.ws.api.stub.PermissionDTO;
+import org.wso2.identity.integration.common.clients.usermgt.remote.RemoteUserStoreManagerServiceClient;
+
+import java.io.IOException;
+import javax.xml.xpath.XPathExpressionException;
+
+public class UserMeNegativeTest extends UserChallengeTest {
+    protected String user;
```
Comment (omindu) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserMeNegativeTest.java]:
```suggestion
    
    protected String user;
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    @AfterMethod(alwaysRun = true)
+    public void testFinish() {
+
+        RestAssured.basePath = StringUtils.EMPTY;
+    }
+
+    protected Response getResponseOfGet(String endpointUri) {
+        return given().auth().preemptive().basic(authenticatingUserName, authenticatingCredential)
```
Comment (omindu) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserChallengeTest.java]:
```suggestion
        
        return given().auth().preemptive().basic(authenticatingUserName, authenticatingCredential)
```

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserMeNegativeTest.java [hunk 14]
```diff
@@ -0,0 +1,185 @@
...
+    protected String role;
+
+    @Factory(dataProvider = "restAPIUserConfigProvider")
+    public UserMeNegativeTest(TestUserMode userMode, String userName, String password, String role, String permission,
+                              boolean isAddUser)
+            throws Exception {
```
Comment (omindu) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/identity/integration/test/rest/api/user/challenge/v1/UserMeNegativeTest.java]:
```suggestion
            throws Exception {
            
```

**Pull request merged**